models/src/main/scala/coop/rchain/casper/protocol/PacketTypeTag.scala

```diff
@@ -0,0 +1,99 @@
+package coop.rchain.casper.protocol
+
+import java.io.IOException
+
+import com.google.protobuf.ByteString
+import coop.rchain.comm.protocol.routing.Packet
+import enumeratum._
+import scalapb.{GeneratedMessage, GeneratedMessageCompanion, Message}
+
+import scala.util.{Failure, Try}
+
+sealed abstract class PacketTypeTag extends EnumEntry
+
+object PacketTypeTag extends Enum[PacketTypeTag] {
+  case object BlockMessage             extends PacketTypeTag
+  case object HasBlockRequest          extends PacketTypeTag
+  case object HasBlock                 extends PacketTypeTag
+  case object BlockRequest             extends PacketTypeTag
+  case object ForkChoiceTipRequest     extends PacketTypeTag
+  case object ApprovedBlock            extends PacketTypeTag
+  case object ApprovedBlockRequest     extends PacketTypeTag
+  case object BlockApproval            extends PacketTypeTag
+  case object UnapprovedBlock          extends PacketTypeTag
+  case object NoApprovedBlockAvailable extends PacketTypeTag
+
+  override val values = findValues
+
+  sealed abstract class ValueOf[Of <: PacketTypeTag] {
```

Comment:
This is also a good candidate for `sealed trait`

Reply:
I developed a habit of using `trait` only when specifically needed. Abstract classes used to compile do better code (no copies of methods), you have no problem with passing parameters, using `val`s etc. But I agree that in this case it might not really matter. For more details please see: https://www.scala-lang.org/blog/2016/07/08/trait-method-performance.html